File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/CountPathsForSum.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class CountPathsForSum {
  static class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
      this.val = val;
      this.left = null;
      this.right = null;
    }
  }

  private static TreeNode buildTree(String[] nodes) {
    return buildTree(0, nodes);
  }

  private static TreeNode buildTree(int currIdx, String[] nodes) {
    if (currIdx >= nodes.length || nodes[currIdx].equalsIgnoreCase("null")) {
      return null;
    }

    TreeNode node = new TreeNode(Integer.parseInt(nodes[currIdx]));
    node.left = buildTree(2 * currIdx + 1, nodes);
    node.right = buildTree(2 * currIdx + 2, nodes);

    return node;
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int sum = in.nextInt();
        in.nextLine();

        String[] nodes = in.nextLine().trim().split(" ");

        TreeNode root = buildTree(nodes);
        System.out.println(countPathsForsum(root, sum));
      }
    }
  }

  static class IntWrapper {
    int ival;

    IntWrapper(int ival) { this.ival = ival; }
  }

  // O(n2) / O(1)
  public static int countPathsForsum(TreeNode root, int sum) {
    IntWrapper totPaths = new IntWrapper(0);
    countPathsForsum(root, sum, totPaths, new ArrayList<>());

    return totPaths.ival;
  }

  private static void countPathsForsum(TreeNode root, int sum,
                                       IntWrapper totPaths,
                                       List<Integer> currPath) {
    if (root == null) {
      return;
    }

    currPath.add(root.val);

    int cnt = 0, currSum = 0;
    for (int i = 0; i < currPath.size() - 1; i++) {
      currSum = 0;
      for (int j = i; j < currPath.size(); j++) {
        currSum += currPath.get(j);

        if (currSum == sum) {
          cnt++;
          break;
        }
      }
    }

    totPaths.ival += cnt;

    countPathsForsum(root.left, sum, totPaths, currPath);
    countPathsForsum(root.right, sum, totPaths, currPath);
    currPath.remove(currPath.size() - 1);
  }
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_3/count_paths_for_sum.cpp
#include <algorithm>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;

  TreeNode(int val) : val(val), left(nullptr), right(nullptr) {}
};

TreeNode *build_tree(int curr_idx, const vector<string> &nodes) {
  if (curr_idx >= nodes.size() || nodes[curr_idx] == "null") {
    return nullptr;
  }

  TreeNode *node = new TreeNode(stoi(nodes[curr_idx]));
  node->left = build_tree(2 * curr_idx + 1, nodes);
  node->right = build_tree(2 * curr_idx + 2, nodes);

  return node;
}

TreeNode *build_tree(const vector<string> &nodes) {
  return build_tree(0, nodes);
}

void count_paths_for_sum(TreeNode *root, int sum, vector<int> &curr_path,
                         int &tot_paths) {
  if (root == nullptr) {
    return;
  }

  curr_path.push_back(root->val);

  int curr_sum = 0;
  for (int i = curr_path.size() - 1; i >= 0; i--) {
    curr_sum += curr_path[i];

    if (curr_sum == sum) {
      tot_paths++;
    }
  }

  count_paths_for_sum(root->left, sum, curr_path, tot_paths);
  count_paths_for_sum(root->right, sum, curr_path, tot_paths);
  curr_path.pop_back();
}

// O(n2) / O(n)
int count_paths_for_sum(TreeNode *root, int sum) {
  vector<int> curr_path;
  int tot_paths = 0;

  count_paths_for_sum(root, sum, curr_path, tot_paths);

  return tot_paths;
}

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int tt, sum;
  string s;

  cin >> tt;
  while (tt--) {
    cin >> sum;
    cin.ignore(1, '\n');

    getline(cin, s);

    stringstream ss(s);
    string tmp;
    vector<string> nodes;

    while (getline(ss, tmp, ' ')) {
      nodes.push_back(tmp);
    }

    TreeNode *root = build_tree(nodes);
    cout << count_paths_for_sum(root, sum) << "\n";
  }

  return 0;
}